clang/lib/Sema/TreeTransform.h

```diff
@@ -7677,27 +7677,23 @@ QualType TreeTransform<Derived>::TransformAttributedType(TypeLocBuilder &TLB,
       switch (TL.getAttrKind()) {
       case attr::PtrSingle:
         AttrName = "__single";
-        LLVM_FALLTHROUGH;
-      case attr::PtrUnsafeIndexable: {
+        break;
+      case attr::PtrUnsafeIndexable:
         AttrName = AttrName ? AttrName : "__unsafe_indexable";
-        QualType ShouldBePtrTy =
-            modifiedType.getDesugaredType(SemaRef.getASTContext());
-        if (!ShouldBePtrTy->isPointerType()) {
-          const auto *Attr = TL.getAttr();
-          // Use the attribute spelling if it's available.
-          // FIXME: When would it not be available?
-          AttrName = Attr ? Attr->getSpelling() : AttrName;
-          SemaRef.Diag(Attr ? Attr->getLocation()
-                            : TL.getModifiedLoc().getBeginLoc(),
-                       diag::err_attribute_pointers_only)
-              << AttrName << 0;
-          return QualType();
-        }
         break;
-      }
       default:
         break; // Nothing to do
       }
+      if (AttrName && !modifiedType->getAs<PointerType>()) {
```

Comment:
I just realised it's not obvious to me why `TreeTransform` is the right place for this diagnostic, rather than an override in `TemplateInstantiator`. The diagnostic above indicates that it actually may be, but it'd be good to test that this correctly handles templated methods inside a templated class, and partial specialisation.